cli/maintainer/src/maintenance.rs

```diff
@@ -754,33 +772,75 @@ impl SolidoState {
-        // to use the actual `spl_token_swap::state::SwapV1` instance, but it is not `Send`, so
-        // we can’t store it in our state.
-        let ust_per_st_sol = Rational {
-            numerator: anker_state.pool_ust_balance.0,
-            denominator: anker_state.pool_st_sol_balance.0,
-        };
-        let expected_proceeds = (rewards * ust_per_st_sol)
-            .map(|x| MicroUst(x.0))
-            .unwrap_or(MicroUst(0));
+        let expected_st_sol_price_in_ust = get_one_st_sol_for_ust_price_from_pool(
+            &anker_state.constant_product_calculator,
+            &anker_state.pool_st_sol_account,
+            &anker_state.pool_ust_account,
+            anker_state.pool_st_sol_balance,
+            anker_state.pool_ust_balance,
+        )
+        .ok()?;
+
+        let expected_proceeds = MicroUst(rewards.0 * expected_st_sol_price_in_ust.0);
         // We want at least 0.01 UST out if we are going to do the swap at all.
         let min_proceeds = MicroUst(10_000);
 
+        // Check if we can sell the rewards with the preset slippage tolerance.
+        // Note that this might change when the instruction gets included in the block.
+        // TODO(#539): Try to arbitrage in case this condition does not hold.
+        let minimum_ust_amount_for_rewards = anker_state
+            .anker
+            .historical_st_sol_prices
+            .minimum_ust_swap_amount(rewards, anker_state.anker.sell_rewards_min_out_bps)
+            .ok()?;
+
         // We should not call the instruction if the rewards are 0, or if the rewards are so small
         // that the transaction cost is a significant portion of the rewards.
-        if rewards < min_rewards_to_sell || expected_proceeds < min_proceeds {
+        if rewards < min_rewards_to_sell {
+            return None;
+        }
+
+        let oldest_price_sample = anker_state.anker.historical_st_sol_prices.first();
+        let slots_elapsed_since_oldest_sample =
+            self.clock.slot.saturating_sub(oldest_price_sample.slot);
+
+        let youngest_sample = anker_state.anker.historical_st_sol_prices.last();
+        let slots_elapsed_since_youngest_sample =
+            self.clock.slot.saturating_sub(youngest_sample.slot);
+
+        // Time to update the historical price
+        if slots_elapsed_since_oldest_sample > POOL_PRICE_MAX_SAMPLE_AGE
+            || oldest_price_sample.slot == 0
+        {
+            if slots_elapsed_since_youngest_sample < POOL_PRICE_MIN_SAMPLE_DISTANCE {
+                None
+            } else {
+                Some(MaintenanceInstruction::new(
+                    anker_state.get_fetch_pool_price_instruction(self.solido_address),
+                    MaintenanceOutput::FetchPoolPrice {
+                        expected_st_sol_price_in_ust,
+                    },
+                ))
+            }
+        } else if expected_proceeds < min_proceeds
+            || expected_proceeds < minimum_ust_amount_for_rewards
```

Comment:
We can add an early return below the definition of `minimum_ust_amount_for_rewards` as well. I don’t have a strong opinion on it, it’s mostly a matter of preference, but I feel it would make the function a bit easier to read.

Reply:
Indeed, much better :)